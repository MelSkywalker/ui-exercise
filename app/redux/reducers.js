export const initialState = {
		emailsById: {},
		emailsIdArray: [],
		tags: [],
		messageCount: null,
		selectedEmails: []
};

// Actions
export const LOAD_EMAIL_LIST = 'LOAD_EMAIL_LIST';
export const SELECT_EMAIL = 'SELECT_EMAIL';
export const DELETE_MESSAGE = 'DELETE_MESSAGE';

// Reducers
const emailsReducer = (state = initialState, action) => {
	const { type, payload } = action;

	switch (type) {
		case LOAD_EMAIL_LIST: {
			const { messages } = payload;
			const emailsById = {};
			let allTags = [];
			messages.forEach((message) => {
				emailsById[message.id] = message; // normalization
				allTags.push(...message.tags);
			});
			const emailsIdArray = messages.map(message => message.id);
			const tagSet = new Set(allTags);
			const tags = Array.from(tagSet);
			const messageCount = emailsIdArray.length;

			return {
				...state,
				emailsById,
				emailsIdArray,
				tags,
				messageCount
			};
		}
		case SELECT_EMAIL: {
			const { selectedItemId, selectedItemStatus } = payload;
			const { selectedEmails } = state;
			const items = selectedEmails;
			if (selectedItemStatus && !items.includes(selectedItemId)) {
				items.push(selectedItemId)
			} else {
				const index = items.indexOf(selectedItemId);
				items.splice(index, 1);
			}
			return {
				...state,
				selectedEmails: items
			};
		}
		case DELETE_MESSAGE: {
			const { emailsById, emailsIdArray, selectedEmails } = state; // emailsIdList
			if (selectedEmails.length) {
				const emailsToRemove = selectedEmails;
				const newEmailsById = { ...emailsById };
				const newIdArray = [...emailsIdArray];

				emailsToRemove.forEach(id => {
					delete newEmailsById[id];
					const index = newIdArray.indexOf(id); // TODO: automate emailsIdArray hydration
					newIdArray.splice(index, 1);
				});

				return {
					...state,
					emailsById: newEmailsById,
					emailsIdArray: newIdArray,
					selectedEmails: []
				}
			}
			return state;
		}
		default:
		return state;
	}
}

export default emailsReducer;
