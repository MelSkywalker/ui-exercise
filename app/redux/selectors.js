export const emailSelector = (state) => {
	return {
		emailsById: state.emailsById,
		emailsIdArray: state.emailsIdArray
	}
}

export const sideBarSelector = (state) => {
	return {
		tagList: state.tags,
		messageCount: state.messageCount
	}
}

