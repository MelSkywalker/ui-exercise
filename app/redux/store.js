import { createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import emailsReducer from './reducers';

const store = createStore(emailsReducer, composeWithDevTools());

export default store;
