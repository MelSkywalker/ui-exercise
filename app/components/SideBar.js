import React from 'react';
import { shallowEqual, useSelector, useDispatch } from 'react-redux';
import { FILTER_EMAILS } from '../redux/reducers';
import { sideBarSelector } from '../redux/selectors';

import Tag from './Tag';
import InboxIcon from './svg/InboxIcon';
import ImportantIcon from './svg/ImportantIcon';

import './sideBar.scss';

const SideBar = () => {
	const { tagList, messageCount } = useSelector(sideBarSelector, shallowEqual);
	const dispatch = useDispatch();

	const handleFilterItems = (e) => {
		const filterToApply = e ? e.target.name : 'all';
		dispatch({
			type: FILTER_EMAILS,
			payload: {
				filterToApply: filterToApply
			}
		})
	}

	return(
		<div className="side-bar">
			<button className="inbox-button side-button" name="all" onClick={handleFilterItems}>
				<InboxIcon className="side-item" />
				<p className="side-item inbox">Inbox</p>
				<span className="message-count">{messageCount}</span>
			</button>
			<button className="side-button">
				<ImportantIcon className="side-item" />
				<p className="side-item">Important</p>
			</button>
			<h3 className="tag-title">Tags:</h3>
			<div className="tags-container">
				{tagList.map((tag) => (
					<Tag key={tag} tagName={tag} handleFilterItems={handleFilterItems} />
				))}
			</div>
		</div>
	);
}

export default SideBar;
