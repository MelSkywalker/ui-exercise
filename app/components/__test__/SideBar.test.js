import React from 'react';
import * as reactRedux from 'react-redux';
import { shallow } from 'enzyme';
import { initialState } from '../../redux/reducers';

import SideBar from '../SideBar';

describe('components | SideBar', () => {
	const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
	const useDispatchSpy = jest.spyOn(reactRedux, 'useDispatch');
	const mockDispatch = jest.fn();
	useDispatchSpy.mockReturnValue(mockDispatch);

	const mockStore = {
		...initialState,
		tagList: ['work', 'travel'],
		messageCount: 3
	};

	beforeEach(() => {
		useSelectorMock.mockClear();
		useDispatchSpy.mockClear();
	});

	it('renders the component', () => {
		useSelectorMock.mockReturnValue(mockStore);
		const component = shallow(<SideBar />);
		expect(component).toMatchSnapshot();
	});

	it('handles click on inbox button', () => {
		useSelectorMock.mockReturnValue(mockStore);
		const component = shallow(<SideBar />);
		component.find('.inbox-button').simulate('click');
		expect(mockDispatch).toHaveBeenCalled();
	});
	it('handles click on tag button', () => {
		useSelectorMock.mockReturnValue(mockStore);
		const component = shallow(<SideBar />);
		console.log(component.find('.tags-container').find('Tag').first().dive().find('button'));
		component.find('.tags-container').find('Tag').first().dive().find('button').simulate('click');
		expect(mockDispatch).toHaveBeenCalled();
	});
});
