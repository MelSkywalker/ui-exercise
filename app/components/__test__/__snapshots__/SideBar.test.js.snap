// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components | SideBar renders the component 1`] = `
<div
  className="side-bar"
>
  <button
    className="inbox-button side-button"
    name="all"
    onClick={[Function]}
  >
    <InboxIcon
      className="side-item"
    />
    <p
      className="side-item inbox"
    >
      Inbox
    </p>
    <span
      className="message-count"
    >
      3
    </span>
  </button>
  <button
    className="side-button"
  >
    <ImportantIcon
      className="side-item"
    />
    <p
      className="side-item"
    >
      Important
    </p>
  </button>
  <h3
    className="tag-title"
  >
    Tags:
  </h3>
  <div
    className="tags-container"
  >
    <Tag
      handleFilterItems={[Function]}
      key="work"
      tagName="work"
    />
    <Tag
      handleFilterItems={[Function]}
      key="travel"
      tagName="travel"
    />
  </div>
</div>
`;
